File: app/src/main/java/io/github/anonymous/pepper_realtime/controller/AudioVolumeController.java
package io.github.anonymous.pepper_realtime.controller;

import android.content.Context;
import android.media.AudioManager;
import android.util.Log;

/**
 * Controls system audio volume.
 */
public class AudioVolumeController {
    private static final String TAG = "AudioVolumeController";

    /**
     * Set media volume as a percentage (0-100)
     */
    public void setVolume(Context context, int percentage) {
        try {
            AudioManager audioManager = (AudioManager) context.getSystemService(Context.AUDIO_SERVICE);
            if (audioManager != null) {
                int maxVol = audioManager.getStreamMaxVolume(AudioManager.STREAM_MUSIC);
                double volume = percentage / 100.0;
                int targetVol = Math.max(0, Math.min(maxVol, (int) Math.round(volume * maxVol)));
                audioManager.setStreamVolume(AudioManager.STREAM_MUSIC, targetVol, 0);
                Log.d(TAG, "Set volume to " + percentage + "% (hardware level: " + targetVol + "/" + maxVol + ")");
            }
        } catch (Exception e) {
            Log.w(TAG, "Setting media volume failed", e);
        }
    }
}

File: app/src/main/java/ch/fhnw/pepper_realtime/controller/AudioVolumeController.kt
package ch.fhnw.pepper_realtime.controller

import android.content.Context
import android.media.AudioManager
import android.util.Log
import kotlin.math.max
import kotlin.math.min
import kotlin.math.roundToInt

/**
 * Controls system audio volume.
 */
class AudioVolumeController {

    companion object {
        private const val TAG = "AudioVolumeController"
    }

    /**
     * Set media volume as a percentage (0-100)
     */
    fun setVolume(context: Context, percentage: Int) {
        try {
            val audioManager = context.getSystemService(Context.AUDIO_SERVICE) as? AudioManager
            if (audioManager != null) {
                val maxVol = audioManager.getStreamMaxVolume(AudioManager.STREAM_MUSIC)
                val volume = percentage / 100.0
                val targetVol = max(0, min(maxVol, (volume * maxVol).roundToInt()))
                audioManager.setStreamVolume(AudioManager.STREAM_MUSIC, targetVol, 0)
                Log.d(TAG, "Set volume to $percentage% (hardware level: $targetVol/$maxVol)")
            }
        } catch (e: Exception) {
            Log.w(TAG, "Setting media volume failed", e)
        }
    }
}

